ui/app/src/state/reducers/pathNavigator.ts

```diff
@@ -0,0 +1,182 @@
+
+export const pathNavigatorFetchPathFailed = createAction('PATH_NAVIGATOR_INIT_FETCH_PATH_FAILED');
+
+export const pathNavigatorSetKeyword = createAction<PayloadWithId<{ keyword: string }>>('PATH_NAVIGATOR_INIT_SET_KEYWORD');
+
+const reducer = createReducer<LookupTable<WidgetState>>(
+  {},
+  {
+    [pathNavigatorInit.type]: (state, { payload: { id, path, locale } }) => {
+      return {
+        ...state,
+        [id]: {
+          rootPath: path,
+          currentPath: path,
+          localeCode: locale,
+          keyword: '',
+          isSelectMode: false,
+          hasClipboard: false,
+          itemsInPath: null,
+          items: {},
+          breadcrumb: [],
+          selectedItems: [],
+          leafs: [],
+          limit: 10,
+          offset: 0,
+          count: 0
+        }
+      };
+    },
+    [pathNavigatorUpdate.type]: (state, { payload: { id, state: updatedState } }) => {
+      return {
+        ...state,
+        [id]: updatedState
+      };
+    },
+    [pathNavigatorFetchPath.type]: (state) => state,
+    [pathNavigatorSetCurrentPath.type]: (state, { payload: { id, path } }) => {
+      return {
+        ...state,
+        [id]: {
+          ...state[id],
+          keyword: '',
+          currentPath: path
+        }
+      };
+    },
+    [pathNavigatorFetchPathComplete.type]: (state, { payload: { id, response } }) => {
+      const path = state[id].currentPath;
+      // Check and handle if the item has no children
+      if (
+        response.length === 0 &&
+        // If it is the root path, we want to show the empty state,
+        // vs child paths, want to show the previous path and inform
+        // that there aren't any items at that path
+        withoutIndex(path) !== withoutIndex(state[id].rootPath)
+      ) {
+        let pieces = path.split('/').slice(0);
+        pieces.pop();
+        if (path.includes('index.xml')) {
+          pieces.pop();
+        }
+        let nextPath = pieces.join('/');
+        if (nou(state[id].items[nextPath])) {
+          nextPath = withIndex(nextPath);
+        }
+        return {
+          ...state,
+          [id]: {
+            ...state[id],
+            // Revert path to previous (parent) path
+            currentPath: nextPath,
+            leafs: state[id].leafs.concat(path)
+          }
+        };
+      } else {
+        const nextItems = {
+          ...state[id].items,
+          [response.parent.id]: response.parent,
+          ...createLookupTable(response)
+        };
+        const widgetState = {
+          ...state[id],
+          breadcrumb: itemsFromPath(path, state[id].rootPath, nextItems),
+          itemsInPath: response.map((item) => item.id),
+          items: nextItems,
+          count: response.length
+        };
+        localStorage.setItem(`craftercms.pathNavigator.${id}`, JSON.stringify(widgetState));
```

Comment:
Reducers should not access local storage. Epics or the component may. 

Don't store data (items). Only store state necessary to restore prior state.